src/dependencies.js

```diff
@@ -0,0 +1,80 @@
+'use strict'
+
+const fs = require('fs-extra')
+const path = require('path')
+const semver = require('semver')
+
+/**
+ * Determine whether GConf is a necessary dependency, given the Electron version.
+ */
+function getGConfDepends (version, distro) {
+  const gconf = distro === 'debian' ? 'libgconf2-4' : 'GConf2'
+  return semver.lt(version, '3.0.0-beta.1') ? [gconf] : []
+}
+
+/**
+ * Determine the GTK dependency based on the Electron version in use.
+ */
+function getGTKDepends (version, distro) {
+  const gtk2 = distro === 'debian' ? 'libgtk2.0-0' : 'gtk2'
+  const gtk3 = distro === 'debian' ? 'libgtk-3-0' : 'gtk3'
+  return semver.gte(version, '2.0.0-beta.1') ? gtk3 : gtk2
+}
+
+/**
+ * Determine the dependencies for the `shell.moveItemToTrash` API, based on the
+ * Electron version in use.
+ */
+function getTrashDepends (version, distro) {
+  const gvfs = distro === 'debian' ? 'gvfs-bin' : 'gvfs-client'
+  const glib = distro === 'debian' ? 'libglib2.0-bin' : 'glib2'
+  if (semver.lt(version, '1.4.1')) {
+    return gvfs
+  } else if (semver.lt(version, '1.7.2')) {
+    return `kde-cli-tools | kde-runtime | trash-cli | ${gvfs}`
+  } else {
+    return `kde-cli-tools | kde-runtime | trash-cli | ${glib} | ${gvfs}`
+  }
+}
+
+/**
+ * Determine whether libuuid is necessary, given the Electron version.
+ */
+function getUUIDDepends (version, distro) {
+  const uuid = distro === 'debian' ? 'libuuid1' : 'libuuid'
+  return semver.gte(version, '4.0.0-beta.1') ? [uuid] : []
+}
+
+/**
+ * Reads the electron version file
+ */
+function getElectronVersion (options) {
+  return fs.readFile(path.resolve(options.src, 'version'))
+    // The content of the version file pre-4.0 is the tag name, e.g. "v1.8.1"
+    // The content of the version file post-4.0 is just the version
+    .then(tag => tag.toString().trim())
+}
+
+module.exports = {
+  getElectronVersion: getElectronVersion,
+  getGConfDepends: getGConfDepends,
+  getGTKDepends: getGTKDepends,
+  getTrashDepends: getTrashDepends,
+  getUUIDDepends: getUUIDDepends,
+
+  /**
+   * Determine the default dependencies for an Electron application.
+   */
+  getDepends: function getDepends (version, distro) {
```

Comment:
I don't think passing in the distro is a very scalable way to do this. It would be a painful experience if someone were to, for example, write an `electron-installer-gentoo` module. Instead of `distro`, the installer module should pass in a `dependency_map` of `dependency: package_name`.

Reply:
I'm not sure if I understand, are you saying that `getDepends` should return a `Map` object containing the names of dependencies in a non-distro specific way, so that each installer module has to translate that to their own distro? Or is there something else I'm missing?

Reply:
I see now, I was proposing returning an array with `[gconf, gtk2, xtst, notify, ...]` and each distro will translate that on their side. That way if a new dependency is added, we return the array with a new item and the installers have to find the appropriate dependency name for it.
If we take a `Map` as input and there's changes in the dependencies and the object is missing that key, it will be a breaking change on the installers side and throw errors here for missing a key. This module will have to release a `major` version update every time that happens to signal to the installers of these changes. Translating on their side, it's only a `minor` change since they can ignore any items they don't recognize, and still be able to work. Dependency changes for new versions of electron should be adding a feature to this module rather than a breaking change.
In either way, none rely on `distro` as input and more importantly, we're going to need something on the README explaining what this module `dependencyMap` should look, either as input or output.

Originally, I was under the impression that this module was going to deal with dependencies completely. If there's any change, the installers don't have to change their code at all, they can rely on getting the exact set of dependencies with their proper names. Also making any dependency change a `minor` release.
I realize that the current code is not portable. Maybe we could have `json` files with the `dependencyMap` inside this repo, one for each distro. If there's a change, we can modify those `json`. If another distro wants to use this module, they can just submit a PR with a new `json` for theirs. Thoughts?

Sorry for the long comment, just wanted to make sure we've thought of all the ways we can implement this and their consequences. In the end, I rely on your expertise and will make whatever change you think is better.

Reply:
> I see now, I was proposing returning an array with `[gconf, gtk2, xtst, notify, ...]` and each distro will translate that on their side. That way if a new dependency is added, we return the array with a new item and the installers have to find the appropriate dependency name for it.

My problem with an array is keeping track of which array position is for which dependency, over time.

> If we take a `Map` as input and there's changes in the dependencies and the object is missing that key, it will be a breaking change on the installers side and throw errors here for missing a key. This module will have to release a `major` version update every time that happens to signal to the installers of these changes. Translating on their side, it's only a `minor` change since they can ignore any items they don't recognize, and still be able to work.

I'm OK with this.

> Maybe we could have `json` files with the `dependencyMap` inside this repo, one for each distro. If there's a change, we can modify those `json`. If another distro wants to use this module, they can just submit a PR with a new `json` for theirs.

I think in terms of encapsulation, I'd rather have the installer modules manage their own package dependencies.

> Sorry for the long comment, just wanted to make sure we've thought of all the ways we can implement this and their consequences.

I appreciate the thought you've put into this.